Fix LightmapGI not working with MultiMeshes

tdaven opened the pull request:
Resolves https://github.com/godotengine/godot/issues/56027

This commit adds support for calling get_meshes on the multimesh_instance_3d.  The order of the transform and mesh differs from get_bake_meshes from gridmap so swap the other so they match.  This appears to fix the issue.

This fixes uses the attempt from https://github.com/godotengine/godot/issues/56027#issuecomment-2192837710 but fixes the ordering in the return so the bake works correctly.

tdaven commented:
An easy alternative way to fix this would be to add a get_bake_mesh that returns the result in the correct order.  I'm not sure what is best.  It would duplicate a lot of get_meshes, which was the primary reason I just added an arg with a default value that preserves the old behavior but allows swapping for this use.

tdaven commented:
It wasn't clear to me if we want to be doing more of the work related to UV2 that is done for normal meshes or not.

tdaven commented:
There is still something wrong.  The multi mesh seems to be getting dynamic shadows where the mesh instances are not.  Will investigate further.

clayjohn commented on scene/3d/lightmap_gi.cpp:
```diff
@@ -354,7 +355,16 @@ void LightmapGI::_find_meshes_and_lights(Node *p_at_node, Vector<MeshesFound> &m
+		if (multi_mesh) {
+			// By default, get_meshes returns an Array with the transform first where the code
+			// below expects the mesh first.  Request the order swapped so baking works.
+			bmeshes = multi_mesh->get_meshes(true);
```
Instead of passing a parameter to get_meshes, I would just swap the order here. Since this function is the one with unique expectations, its better to keep the complexity here, instead of spreading it to MultiMeshInstance

clayjohn commented:
> There is still something wrong. The multi mesh seems to be getting dynamic shadows where the mesh instances are not. Will investigate further.

There is code to register the Lightmap with a MeshInstance when the Lightmap is loaded. That is probably missing MultiMeshInstances. I would start by looking here and making sure the lightmap is registered properly:

tdaven commented:
> > There is still something wrong. The multi mesh seems to be getting dynamic shadows where the mesh instances are not. Will investigate further.
> 
> There is code to register the Lightmap with a MeshInstance when the Lightmap is loaded. That is probably missing MultiMeshInstances. I would start by looking here and making sure the lightmap is registered properly:
> 
> https://github.com/godotengine/godot/blob/5ccbf6e4c794a4e47456edd9434b75fcd6096a2f/scene/3d/lightmap_gi.cpp#L1386

Yeah, doing this has stopped the dynamic shadows but still not working entirely.  I think is a problem where its not using the lightmap possibly because something is missing to tell it to use UV2.  Still investigating what is going wrong.  I tried looking at gridmap but it seems entirely broken for lightmap_gi as well for similar reasons.

tdaven commented:
This message was pasted to the rocket chat thread.  Pasting here in-case its helpful for others.

I did a bit of investigation with gridmaps and I don't think this can be implemented like gridmaps. Gridmaps do have some uses of multimesh in the code, but it hardly used at all, especially for lightmaps. Everytime i've followed how grid map does it, you just get duplicate geometry. This is because gridmap spawns meshs when you bake the meshes.

I have a branch which does not use multimesh at all here ([gridmap-no-multimesh](https://github.com/godotengine/godot/compare/master...tdaven:godot:gridmap-no-multimesh)). This branch won't display anything when editted, until you bake lightmaps. When it makes the baked meshes, it spawns new meshes. You can see the place this mostly happens here([gridmap mesh instancing](https://github.com/godotengine/godot/blob/master/modules/gridmap/grid_map.cpp#L1294)). It also does some of this elsewhere to update. Its unclear to me where the multimesh stops drawing but it does. This was further confirmed by instrumenting the code and using the debugger. In the state mentioned above where the grid doesn't have baked meshes(and it doesn't show things properly), the multimesh is used and you can see it being drawed(assuming you're not on my no multimesh branch), but as soon as you bake meshes, it stops and instead you are only drawing mesh(s) made during baking linked above.

Here is an image from my no multi-mesh branch showing that it still renders and with working lightmaps:
![image](https://github.com/user-attachments/assets/6dbc759a-34cc-4f03-95cb-0cbcbd659af1)

tdaven commented:
This my latest push I have basic support implemented for lightmapgi and multimesh.  Its not complete though and there are number of things I plan to revisit.  But it does work across all three renderers and produces identical results.

Small test scene.  One side is made with individual mesh instances, the other is multi-mesh.  They look identical as desired.  Multimesh is on the left in case you are wondering.
![image](https://github.com/user-attachments/assets/d4b52afe-0e54-4634-96e3-50261416183e)

tdaven committed:
Implement lightmap support for multimesh.

tdaven commented:
I think this is more or less ready for more wide spread review.  Lightmap support for mulit-meshes is implemented for compatibility, mobile and forward+.  

In order to implement this support, I've added a way to have per instance lightmap uv scale and slice index.  For the most part, this has just been appended to the existing instances buffer.  This isn't the only way to implement it.  At one point I had mobile/forward+ using their own buffer for only this data.  Its mostly tradeoff between some complexity in dealing with one buffer vs the complexity of adding a new buffer to a uniform set.

There are likely a few debug prints remaining I haven't removed yet.  The compatibility support has to use another vertex attribute binding.  I just used 2 which was documented as unused in the shader.  This was helpful since it kept the max under 16 which is the limit for some systems (I think webgl2 and llvmpipe for example).

One question was can we implement it like gridmap does.  I don't think so.  Gridmap seems to just spawn normal mesh instances which defeats the benefit multi-mesh(but where there is already batching maybe its not needed much).  Presumably, once this merges, gridmap can be fixed to just use multi-mesh more thoroughly and not have to implement its own lightmap support (but I've not investigated what all that would take yet).

Calinou reviewed (commented):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/9e6098432aac35bae42c9089a29ba2a80320d823), there are some issues.

The MultiMesh is now taken into account during baking, but the UV mapping on its faces look broken with the MRP from https://github.com/godotengine/godot/issues/56027 (MultiMesh on the right):

![image](https://github.com/user-attachments/assets/c1714f86-2624-4c42-8d1f-0790a0dcbfe1)

tdaven commented:
@Calinou, I think I have fixed the issue you hit plus a few others as well (motion vectors where not being handle properly after I had cleaned up code).

![image](https://github.com/user-attachments/assets/8daee44e-f86e-42f6-8900-44000ac961ae)

I had thought those dark cubes were a problem but I believe they just have basically a negative scale so they are inside out.  Non-lightmap lighting is also dark on them.

You'll want to re-import the the MRP project though.  The main defect was a buffer copy which still had an offset (from one of the way I tried implement this) which resulted in no the entire buffer being copied which may have resulted in the project not saving the transforms correctly.

I also added menu items to view uv1/uv2 plus unwrap for lightmap since it was frustrating not having them.  The code is mostly copy pasted from mesh_instance_3d.  Not idea but it definitely improves the work flow.  This bit could always be strip out to its own PR or maybe some common class both could use.

Calinou commented on editor/plugins/mesh_editor_uv_tools.cpp:
```diff
@@ -0,0 +1,293 @@
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "mesh_editor_uv_tools.h"
+
+#include "editor/editor_node.h"
+#include "editor/editor_string_names.h"
+#include "editor/editor_undo_redo_manager.h"
+// #include "editor/multi_node_edit.h"
+// #include "editor/plugins/node_3d_editor_plugin.h"
+#include "editor/themes/editor_scale.h"
+// #include "scene/3d/navigation_region_3d.h"
+// #include "scene/3d/physics/collision_shape_3d.h"
+// #include "scene/3d/physics/physics_body_3d.h"
+// #include "scene/3d/physics/static_body_3d.h"
+#include "scene/gui/aspect_ratio_container.h"
+// #include "scene/gui/box_container.h"
+#include "scene/gui/dialogs.h"
+// #include "scene/gui/menu_button.h"
+// #include "scene/gui/spin_box.h"
+// #include "scene/resources/3d/concave_polygon_shape_3d.h"
+// #include "scene/resources/3d/convex_polygon_shape_3d.h"
+#include "scene/resources/3d/primitive_meshes.h"
```
Remove all commented out lines here.

Calinou commented on editor/plugins/mesh_editor_uv_tools.cpp:
```diff
@@ -0,0 +1,293 @@
+	}
+};
+
+void MeshEditorUvTools::create_uv_lines(Ref<Mesh> mesh, int p_layer) {
```
The class should be named MeshEditorUVTools (UV instead of Uv), as per Godot's naming conventions.

Calinou reviewed (commented):
Tested locally, it works as expected in all rendering methods. Code looks good to me at a glance (other than the style changes I requested above).

michaelharmonart commented:
super excited for this! awesome work

Repiteo commented:
Needs rebase.

michaelharmonart commented:
@Repiteo are there any specific things you're hoping to merge before this makes it in/any foreseeable conflicts or blockers for merging this?

I know it's not good form to ask to expedite changes, but I'm really hoping this makes it into mainline Godot soon, it looks like @tdaven has done some really great work, and it'll be a huge improvement for my project.

tdaven commented:
I keep occasionally rebasing or can do as needed if someone pings me but after a few months of waiting I kinda stopped regular updates since it just wasted build resources waiting for people to review.

dog-on-moon commented on drivers/gles3/rasterizer_scene_gles3.cpp:
```diff
@@ -3571,6 +3576,7 @@ void RasterizerSceneGLES3::_render_list_template(RenderListParameters *p_params,
 
 				GLuint instance_buffer = 0;
 				uint32_t stride = 0;
+				uint32_t instance_count = mesh_storage->multimesh_get_instance_count(inst->data->base);
```
doing some testing with this PR, an active GPUParticles3D node (compatability) spams the following error:

`ERROR: drivers\gles3\storage\mesh_storage.cpp:1571 - Parameter "multimesh" is null.`

not 100% sure, but i suspect this to be related. also might be a false positive on my part, since i cherry picked it to a custom build

dog-on-moon commented on drivers/gles3/shaders/scene.glsl:
```diff
@@ -2080,9 +2117,7 @@ void main() {
-		vec3 uvw;
-		uvw.xy = uv2 * lightmap_uv_scale.zw + lightmap_uv_scale.xy;
-		uvw.z = float(lightmap_slice);
+		vec3 uvw = lightmap_uvw;
```
i believe this fix will also need to be applied to line 2365

tdaven commented:
Alright, hopefully both issues are now resolved and rebased on master again.

dog-on-moon commented:
Did a little more testing!

- baking a scene with no mesh UV2s, but a MultiMeshInstance3D, will freeze the lightmap baking on "Determining optimal atlas size" -- basically, it seems that the check to popup the "No meshses with lightmapping support to bake." window is ignored whenever a multimesh is involved (even when its mesh has no UV2). this should be fixed

- very unpredictably, one of my lightmaps seem to hard crash on load when there are multiple multimeshes. I was able to capture a traceback of one of these:
![WindowsTerminal_Xebdafb1ax](https://github.com/user-attachments/assets/91e4d71a-a78f-4d54-b516-194f4799eae4)

unfortunately, I could not capture an MRP -- attempting to port the bugged scene over to a separate project led to the multimesh buffer performing a beautiful waltz, incurring a grand frustration within me:

https://github.com/user-attachments/assets/b7748979-c7d0-423a-a38c-23bd9d756dce

I think this crash is caused by some memory issues with multimeshes (especially since I was also having other tracebacks caused by an AnimationTree in the scene attempting to allocate memory). i cannot find any consistency in this behavior, either

I would take a closer look over all of the memory allocations in this PR. if they all seem fine, then I think this memory issue is unrelated -- the PR is good, and the lightmap multimeshes are stable from my testing (whenever the multimeshes themselves are behaving nicely)

tdaven commented:
> baking a scene with no mesh UV2s, but a MultiMeshInstance3D, will freeze the lightmap baking on "Determining optimal atlas size" -- basically, it seems that the check to popup the "No meshses with lightmapping support to bake." window is ignored whenever a multimesh is involved (even when its mesh has no UV2). this should be fixed

I've pushed a fix for this issue.  Was definitely missing something there.

> very unpredictably, one of my lightmaps seem to hard crash on load when there are multiple multimeshes. I was able to capture a traceback of one of these.

Not sure about this one.  Its not crashing in anything related to lightmaps but insteads the gizmos.   Maybe I can try an asan build and see if anything happens but I've not been able to reproduce this one yet with any of my test projects.  I modified on to have multiple multi meshes but haven't gotten a crash yet.

dog-on-moon commented:
Good news, it seems that my memory issues with lightmaps are related to #75485 instead (a persistent memory corruption issue with lightmaps). Calling `multimesh.buffer = PackedFloat32Array()` after modifying instance count appears to be a stable workaround, as mentioned in that issue

this continues to work great after the fixes

blueskythlikesclouds commented on servers/rendering/renderer_rd/shaders/forward_clustered/scene_forward_clustered.glsl:
Not sure that I like regular meshes paying the cost of an extra interpolator just for multimeshes.

Could you not fetch the per-multimesh lightmap UV scale/offset in the fragment shader instead? The transforms buffer is accessible there, IIRC.

tdaven replied:
you can access the ssbo in the fragment shader but you don't have the index to get the data I don't think.  But...we could pass more info the fragment as flat so it doesn't have to interpolate.

Now, whether that is faster or not....i'm less sure.  Its one buffer load in the vertex shader vs doing something on every fragment.  Its already  accessing that ssbo in the fragment lots anyway so its probably all in cache.

tdaven commented:
I do have an updated version of this change.  This broke a lot after the motion vectors merged for compatiblity.  I'll try to get the updated version pushed soon.

TheHyper-Dev commented:
Wonder if this will ever get merged, it's a game changer and you guys are great for working on this

SpockBauru commented:
Just want to add that this is specially important for the Mobile renderer, since there's no auto batching and instancing like Forward+. This means that using MultiMeshes is the only way to use GPU instancing on Mobile.
Currently there's no way to use both GPU Instancing and LightmapGI on Mobile, which is the platform that would most benefit from them.
